src/analyses/variable-sensitivity/abstract_object.cpp

```diff
@@ -0,0 +1,299 @@
+
+abstract_objectt::abstract_objectt(const constant_exprt &expr):
+type(expr.type()), top(true), bottom(false)
+{}
+
+abstract_objectt::~abstract_objectt()
+{
+
+}
+
+const typet &abstract_objectt::get_type() const
+{
+  return type;
+}
+
+/*******************************************************************\
+
+Function: abstract_objectt::merge_state
+
+  Inputs:
+   op1 - the first abstract object
+   op2 - the second abstract object
+
+ Outputs:
+
+ Purpose: Set this abstract object to be the result of merging two
+          other abstract objects. This is the worst case - we can
+          only set to top or bottom.
+
+\*******************************************************************/
+
+bool abstract_objectt::merge_state(
+  const abstract_object_pointert op1, const abstract_object_pointert op2)
+{
+  top=op1->top||op2->top;
+  bottom=op1->bottom && op2->bottom;
+
+  assert(!(top && bottom));
+  return top!=op1->top || bottom!=op1->bottom;
+}
+
+/*******************************************************************\
+
+Function: abstract_objectt::merge
+
+  Inputs:
+   op - the abstract object to merge with
+
+ Outputs:
+
+ Purpose: Set this abstract object to be the result of merging this
+          abstract object and the provided one. See merge_state.
+
+\*******************************************************************/
+
+abstract_object_pointert abstract_objectt::merge(
+  const abstract_object_pointert op, bool &out_any_modifications)
+{
+  assert(type==op->type);
+  abstract_object_pointert m=abstract_object_pointert(new abstract_objectt(*this));
+  out_any_modifications=m->merge_state(abstract_object_pointert(this), op);
+  return m;
+}
+
+/*******************************************************************\
+
+Function: abstract_objectt::expression_transform
+
+  Inputs:
+   expr - the expression to evaluate and find the result of it. this will
+          be the symbol referred to be op0()
+
+ Outputs: Returns the abstract_object representing the result of this expression
+          to the maximum precision available.
+
+ Purpose: To try and resolve different expressions with the maximum level
+          of precision available.
+
+\*******************************************************************/
+
+abstract_object_pointert abstract_objectt::expression_transform(
+  const exprt &expr,
+  const abstract_environmentt &environment,
+  const namespacet &ns) const
+{
+  exprt constant_replaced_expr=expr;
+  constant_replaced_expr.operands().clear();
+  for(const exprt &op : expr.operands())
+  {
+    abstract_object_pointert lhs_abstract_object=environment.eval(op, ns);
+    const exprt &lhs_value=lhs_abstract_object->to_constant();
+
+    if(lhs_value.id()==ID_nil)
+    {
+      // One of the values is true so we can't really do anything more with
+      // this expression and should just return top for the result
+      return abstract_object_pointert(
+        new abstract_objectt(expr.type(), true, false));
```

Comment:
How/why is this different from the `environment.abstract_object_factory(expr.type(), true, false);` ?